tensorboard/webapp/widgets/custom_modal/custom_modal.ts

```diff
@@ -0,0 +1,221 @@
+ * ```
+ *
+ * # Usage
+ * Define a modal using an ng-template:
+ * ```
+ * <ng-template #myModalTemplate>
+ *    <my-awesome-modal-content
+ *      [input1]="input1"
+ *      (output1)="output1"
+ *    >
+ *    </my-awesome-modal-content>
+ * </ng-template>
+ * ```
+ *
+ * Define a ViewChild to reference the template in the component file:
+ * ```
+ * // my_component.ts
+ * ...
+ * @ViewChild('myModalTemplate', {read: TemplateRef})
+ * myModalTemplate!: TemplateRef<unknown>;
+ * ...
+ * ```
+ *
+ * Inject CustomModal into the component
+ * ```
+ * // my_component.ts
+ * ...
+ * constructor(private readonly customModal: CustomModal) {}
+ * ...
+ * ```
+ *
+ * To create a modal, call createNextToElement():
+ * ```
+ * // my_component.ts
+ * ...
+ * onSomeButtonClick(mouseEvent: MouseEventj) {
+ *   this.customModal.createNextToElement(this.myModalTemplate, mouseEvent.target as HTMLElement);
+ * }
+ * ...
+ * ```
+ *
+ * # Testing
+ * A similar setup is required for testing modal behavior from another component.
+ *
+ * Define a separate root component to hold the modal container and ref:
+ * ```
+ * @Component({
+ *   ...,
+ *   template: `<actual-component-to-test></actual-component-to-test>`
+ * })
+ * class TestableComponent {
+ *   constructor(readonly vcRef: ViewContainerRef) {}
+ *   ...
+ * }
+ * ```
+ *
+ * Before each test (e.g. in beforeEach), make sure to add the TestableComponent
+ * to the app component list:
+ * ```
+ * ...
+ * const fixture = TestBed.createComponent(TestableComponent);
+ * const appRef = TestBed.inject(ApplicationRef);
+ * appRef.components.push(fixture.componentRef);
+ * ```
+ */
+
+export class CustomModalRef {
+  overlayRef: OverlayRef;
+  subscriptions: Subscription[];
+
+  constructor(overlayRef: OverlayRef, subscriptions: Subscription[] = []) {
+    this.overlayRef = overlayRef;
+    this.subscriptions = subscriptions;
+  }
+}
+
+@Injectable({providedIn: 'root'})
+export class CustomModal {
+  private customModalRefs: CustomModalRef[] = [];
+
+  constructor(
+    private readonly appRef: ApplicationRef,
+    private readonly overlay: Overlay
+  ) {}
+
+  /** Gets the ViewContainerRef of the app's root component if it exists. */
+  private getModalViewContainerRef(): ViewContainerRef | undefined {
+    const appComponents = this.appRef.components;
+    if (appComponents.length === 0) {
+      // appComponents can potentially be empty in tests.
+      return;
+    }
+
+    const appInstance = appComponents[0].instance;
+    let viewContainerRef: ViewContainerRef = appInstance.vcRef;
+    if (!viewContainerRef) {
+      console.warn(
+        'For proper custom modal function, an ViewContainerRef named `vcRef` is required in the root component.'
+      );
+      return;
+    }
+    return viewContainerRef;
+  }
+
+  /** Creates a modal from a template next to an element. */
+  createNextToElement(
+    templateRef: TemplateRef<unknown>,
+    element: Element,
+    connectedPosition: ConnectedPosition = {
```

Comment:
I don't see where you ever override connectedPosition. I think you're always relying on the default? Could you remove this as an argument to the function and instead make it a class or file-level constant?

Reply:
It's not being used at the moment, but it seems very useful to allow future callers to configure the modal position rather than always create it at the bottom right. It also suits the theme of the function: we can create a modal next to an element, but "next to" how?

I don't have a strong opinion on this though - I can clean it up if you think making this constant is cleaner for now.